discovery-provider/src/tasks/index.py

```diff
@@ -243,6 +248,101 @@ def index_blocks(self, db, blocks_list):
+
+        track_listens = {}
+        try:
+            resp = requests.get(identity_tracks_endpoint, params=params)
+            track_listens = resp.json()
+        except Exception as e:
+            logger.error(
+                f'Error retrieving track play counts - {identity_tracks_endpoint}, {e}'
+            )
+
+        # Insert a new row for each count instance in the plays table
+        plays = []
+        if 'listens' in track_listens:
+            for listen in track_listens['listens']:
+                if 'userId' in listen and listen['userId'] != None:
+                    # If the userId is present, query for exist plays and only
+                    # insert new plays for the difference
+                    user_track_play_count = session.query(
+                        func.count(Play.play_item_id)
+                    ).filter(
+                        Play.play_item_id == listen['trackId'],
+                        Play.user_id == listen['userId']
+                    ).scalar()
+                    new_play_count = listen['count'] - user_track_play_count
+                    if new_play_count > 0:
+                        plays.extend([
+                            Play(
+                                user_id=listen['userId'],
+                                play_item_id=listen['trackId'],
+                                created_at=listen['createdAt'],
+                            ) for _ in range(new_play_count)
+                        ])
+                else:
+                    # For anon track plays, check the current hour play counts
+                    # and only insert new plays for the difference
+                    current_hour_query = dateutil.parser.parse(
+                            listen['createdAt']
+                        ).replace(microsecond=0, second=0, minute=0)
+                    anon_hr_track_play_count = session.query(
+                        func.count(Play.play_item_id)
+                    ).filter(
+                        Play.play_item_id == listen['trackId'],
+                        Play.user_id == None,
+                        Play.created_at >= current_hour_query
+                    ).scalar()
+                    new_play_count = listen['count'] - anon_hr_track_play_count
+                    if new_play_count > 0:
+                        plays.extend([
+                            Play(
+                                play_item_id=listen['trackId'],
+                                created_at=listen['createdAt'],
+                            ) for _ in range(new_play_count)
+                        ])
+
+        if len(plays) > 0:
+            session.bulk_save_objects(plays)
+            session.execute("REFRESH MATERIALIZED VIEW aggregate_plays")
+
 # transactions are reverted in reverse dependency order (social features --> playlists --> tracks --> users)
 def revert_blocks(self, db, revert_blocks_list):
```

Comment:
is there any scenario where we have to revert a play count operation to a point in time?

Reply:
Currently, there isn't but I image when it's indexed from chain we'll have to add revert logic.